src/main/java/com/basho/riak/client/api/commands/ChunkedResponseIterator.java

```diff
@@ -0,0 +1,122 @@
+package com.basho.riak.client.api.commands;
+
+import com.basho.riak.client.core.StreamingRiakFuture;
+import com.basho.riak.client.core.util.BinaryValue;
+
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TransferQueue;
+import java.util.function.Function;
+
+public class ChunkedResponseIterator<FinalT, ChunkT extends Iterable<CoreT>, CoreT> implements Iterator<FinalT>
+{
+    private final int timeout;
+    private volatile BinaryValue continuation = null;
+    private final StreamingRiakFuture<ChunkT, ?> coreFuture;
+    private final TransferQueue<ChunkT> chunkQueue;
+    private final Function<CoreT, FinalT> createNext;
+    private final Function<ChunkT, Iterator<CoreT>> getNextIterator;
+    private final Function<ChunkT, BinaryValue> getContinuationFn;
+
+    private Iterator<CoreT> currentIterator = null;
+
+    public ChunkedResponseIterator(StreamingRiakFuture<ChunkT, ?> coreFuture,
+                                   int pollTimeout,
+                                   Function<CoreT, FinalT> createNextFn,
+                                   Function<ChunkT, Iterator<CoreT>> getNextIteratorFn)
+    {
+        this(coreFuture, pollTimeout, createNextFn, getNextIteratorFn, (x) -> null);
+    }
+
+    public ChunkedResponseIterator(StreamingRiakFuture<ChunkT, ?> coreFuture,
+                                   int pollTimeout,
+                                   Function<CoreT, FinalT> createNextFn,
+                                   Function<ChunkT, Iterator<CoreT>> getNextIteratorFn,
+                                   Function<ChunkT, BinaryValue> getContinuationFn)
+    {
+        this.timeout = pollTimeout;
+        this.coreFuture = coreFuture;
+        this.chunkQueue = coreFuture.getResultsQueue();
+        this.createNext = createNextFn;
+        this.getNextIterator = getNextIteratorFn;
+        this.getContinuationFn = getContinuationFn;
+        loadNextChunkIterator();
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        if (currentIteratorHasNext())
+        {
+            return true;
+        }
+
+        loadNextChunkIterator();
+
+        return currentIteratorHasNext();
+    }
+
+    private boolean currentIteratorHasNext()
+    {
+        return currentIterator != null && currentIterator.hasNext();
+    }
+
+    private boolean possibleChunksRemaining()
+    {
+        // Chunks may remain if :
+        // Core Operation Not Done OR items still in chunk Queue
+        return !coreFuture.isDone() || !chunkQueue.isEmpty();
+    }
+
+    public boolean hasContinuation()
+    {
+        return continuation != null || possibleChunksRemaining();
+    }
+
+    public BinaryValue getContinuation()
+    {
+        return continuation;
+    }
+
+    @Override
+    public synchronized FinalT next()
+    {
+        return createNext.apply(currentIterator.next());
```

Comment:
Lack of synchronization on hasNext() whereas next() is synchronized looks looks very suspicious for me.